Feature/persistencia

morenavic opened the pull request.

morenavic committed:
feat: Persistencia de los datos.

- Se modificó el nombre de application.template-properties por application.properties para que levante la app correctamente y se agregaron los valores necesarios.
- Se creó la clase CrearCursoRepositorioImpl para implementar los métodos de la interfaz ICrearCursoRepositorio.
- Se creó la entidad de Curso con las anotaciones correspondientes.
- Se creó la interfaz del CRUD.
- Se actualizó el pom.xml y el AdapterApplication.

morenavic committed:
feat: se implementaron los repositorios, controller, DTO y configuración para gestión de cursos.

morenavic committed:
feat: se implementó el manejo de excepciones.

morenavic committed:
feat: correcciones en la búsqueda de cursos por fecha.

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/config/CursoConfig.java:
```diff
@@ -0,0 +1,46 @@
+public class CursoConfig {
+
+    //Bean que devuelve la implementación del repositorio de creación de cursos
+    @Bean
```
No hace falta instanciar manualmente los repositorios, ya que la clase CrearCursoRepositorioImpl es un bean gestionado por Spring, y es una implementación de la interfaz ICrearCursoRepositorio

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/CursoController.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Autowired
+    //Inyección de dependencias
+    public CursoController (ICrearCursoInput iCrearCursoInput, IBuscarCursoInput iBuscarCursoInput){
```
Este controller podría separarse en dos (Crear y Buscar), asi tenemos alineado con los RepoImplementation y los usecases. Es sugerencia, no obligatoria

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/CursoController.java:
```diff
@@ -0,0 +1,108 @@
+        );
+
+        //Llamamos al caso de uso
+        boolean cursoCreado = iCrearCursoInput.crearCurso(curso);
```
Que sucede si esto es False?

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/CursoController.java:
```diff
@@ -0,0 +1,108 @@
+    @PostMapping
+    public ResponseEntity<String> crearCursoDTO(@RequestBody CursoDTO cursoDTO){
+        //Convertimos el DTO directamente al modelo del core (Curso)
+        Curso curso = new Curso(
```
Ojo aquí. El constructor de Curso no estaba privado? Tendríamos que llamar al método Instancia, para asegurarnos de que se ejecutan las validaciones del core)

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/CursoController.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    //Metodo privado para convertir un objeto Curso en un objeto CursoDTO.
+    private CursoDTO convertirCursoACursoDTO(Curso curso) {
```
Sugerencia: Para evitar tener que replicar este metodo en distintos lados, podría llevarse a la clase DTO, y que ésta tenga la facultad de convertir el curso, desde y hacia Core

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/dto/CursoDTO.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class CursoDTO {
+
+    private String nombre;
```
Acá, investiva el uso de la librería Jackson, para usar la anotación @JsonProperty

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/exception/ManejadorExcepciones.java:
```diff
@@ -0,0 +1,52 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class ManejadorExcepciones {
```
Bien por el uso del Advice 👏

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/exception/ManejadorExcepciones.java:
```diff
@@ -0,0 +1,52 @@
+        return ResponseEntity.status(HttpStatus.CONFLICT).body(ex.getMessage());
+    }
+
+    @ExceptionHandler(FechaAnteriorException.class)
+    public ResponseEntity<String> manejarFechaAnterior(FechaAnteriorException ex) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(ex.getMessage());
+    }
+
+    @ExceptionHandler(FechaRequeridaException.class)
+    public ResponseEntity<String> manejarFechaRequerida(FechaRequeridaException ex) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(ex.getMessage());
+    }
```
Podés agrupar las excepciones en un solo método, si la respuesta es la misma

@ExceptionHandler({Ex1.class, Ex2.class, Ex3.class})

jruitti commented on adapter/src/main/java/ar/edu/undec/adapter/controller/dto/CursoDTO.java:
```diff
@@ -0,0 +1,50 @@
+        return nombre;
+    }
+
+    public void setNombre(String nombre) {
```
No es recomendable usar Setters. Lo ideal, por inmutabilidad de los objetos, es setear los valores a través de constructor o factoryMethod, y no modificarlos durante el procesamiento del caso de uso

jruitti commented on adapter/src/main/java/repository/BuscarCursoRepositorioImpl.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    //Metodo privado para convertir una colección de entidades a modelos
+    private Collection<Curso> convertirAColeccionDeModelo(Collection<CursoEntity> cursoEntities) {
```
Similar a DTO, este metodo puede ser parte de la case CursoEntity, y que ahi se convierta de un tipo a otro.
También se podría crear una clase Mapper, que tenga los métodos de conversión.
Estas sugerencias son para evitar tener que repetir este código en distintos RepoImplementation

jruitti commented on adapter/src/main/java/repository/CrearCursoRepositorioImpl.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    //Guarda un curso en la BD y devuelve true si se guardó correctamente
+    @Override
```
Acá se puede enriquecer con una anotación @Transactional.

https://docs-spring-io.translate.goog/spring-data/jpa/reference/jpa/transactions.html?_x_tr_sl=en&_x_tr_tl=es&_x_tr_hl=es&_x_tr_pto=tc

jruitti commented on core/src/main/java/curso/modelo/Curso.java:
```diff
@@ -1,34 +1,43 @@
     private Nivel nivel;
 
-    private Curso(String nombre, LocalDateTime fechaCierreInscripcion, Nivel nivel) {
+    public Curso(String nombre, LocalDateTime fechaCierreInscripcion, Nivel nivel) {
```
El constructor debería ser privado, y comunicarse con el método Instancia

jruitti commented on core/src/main/java/curso/usecase/CrearCursoUseCase.java:
```diff
@@ -16,12 +16,16 @@ public CrearCursoUseCase(ICrearCursoRepositorio iCrearCursoRepositorio) {
 
     @Override
     public boolean crearCurso(Curso curso) {
+
+        //Crea una instancia del curso para realizar validaciones
+        Curso cursoInstancia = Curso.instancia(curso.getNombre(),curso.getFechaCierreInscripcion(),curso.getNivel());
+
```
Si el argumento recibido en este método ya es un curso, significa que pasó por alguna etapa de validación. Nunca está de mas ser precavido, pero la pregunta es: ¿Por qué podría llegarnos como parámetro un curso sin validar?

morenavic committed:
Correcciones y mejoras en el código.
- Se eliminaron los @Bean de los repositorios en CursoConfig y se movió la carpeta repository que contiene las implementaciones dentro del paquete escaneado para que Spring las detecte.
- CursoController se separó en dos (Crear y Buscar).
- Se movió lógica de conversión al DTO, se eliminó setters y constructor privado, y se agregó la anotación @JsonProperty.
- Se agruparon excepciones con la misma respuesta.
- Se creó CursoMapper con métodos estáticos para evitar la repetición de código.
- Se aplicó @Transactional en el repositorio.
- Se corrigió el constructor de Curso.
- Se usó Curso.instancia para hacer las validaciones correspondientes.
